src/service/layers-impl.js

```diff
@@ -0,0 +1,634 @@
+    const layouts = this.layouts_;
+    for (let i = 0; i < layouts.length; i++) {
+      const layout = layouts[i];
+      layout.undeclareLayer();
+      layout.forgetParentLayer();
+    }
+  }
+
+  scrolled_(element) {
+    let layout = LayoutElement.forOptional(element);
+    if (layout && layout.isLayer()) {
+      layout.requestScrollRemeasure();
+    } else {
+      layout = this.declareLayer_(element);
+    }
+
+    if (this.onScroll_) {
+      this.onScroll_(/* layer.getElements() */);
+    }
+  }
+
+  onScroll(handler) {
+    this.onScroll_ = handler;
+  }
+}
+
+export class LayoutElement {
+  constructor(element) {
+    element[LAYOUT_PROP] = this;
+    /**
+     * @private @const {!Element}
+     */
+    this.element_ = element;
+
+    /** @type {?LayoutElement|undefined} */
+    this.parentLayer_ = undefined;
+
+    this.needsRemeasure_ = true;
+    this.size_ = SizeWh(0, 0);
+    this.position_ = PositionLt(0, 0);
+
+    this.isLayer_ = false;
+    this.isRootLayer_ = false;
+    this.needsScrollRemeasure_ = false;
+    this.scrollLeft_ = 0;
+    this.scrollTop_ = 0;
+
+    this.children_ = [];
+  }
+
+  /**
+   * Gets the element's LayoutElement instance.
+   * @param {!Element}
+   * @return {!LayoutElement}
+   */
+  static for(element) {
+    return dev().assert(LayoutElement.forOptional(element));
+  }
+
+  /**
+   * Gets the element's LayoutElement instance, if the element has one.
+   * @param {!Element}
+   * @return {?LayoutElement}
+   */
+  static forOptional(element) {
+    return element[LAYOUT_PROP];
+  }
+
+  /**
+   * Finds the element's parent layer
+   * If the element is itself a layer, it returns the layer's parent layer.
+   * @param {!Element}
+   * @param {opt_force=} Whether to force a re-lookup
+   * @return {?LayoutElement}
+   */
+  static getParentLayer(element, opt_force) {
+    if (!opt_force) {
+      const layout = LayoutElement.forOptional(element);
+      if (layout) {
+        return layout.getParentLayer();
+      }
+    }
+
+    const win = element.ownerDocument.defaultView;
+    if (computedStyle(win, element).position == 'fixed') {
+      LayoutLayers.declareLayer(element);
+      return null;
+    }
+
+    let last = element;
+    let op = element.offsetParent;
+    for (let el = last.parentNode; el; el = el.parentNode) {
+      const layout = LayoutElement.forOptional(el);
+      if (layout && layout.isLayer()) {
+        return layout;
+      }
+
+      // now check to see if offsetParent is a fixed layer
+      if (el === op) {
+        if (computedStyle(win, op).position == 'fixed') {
+          LayoutLayers.declareLayer(op);
+          return op;
+        }
+        op = op.offsetParent;
+      }
+
+      last = el;
+    }
+
+
+    if (last.nodeType !== Node.DOCUMENT_NODE) {
+      throw dev().createError('element is not in the DOM tree');
+    }
+    return null;
+  }
+
+  contains(element) {
+    const root = this.element_;
+    return root !== element && root.contains(element);
```

Comment:
This may or may not work for shadow roots (need to check) and definitely wouldn't work for FIE. This could be totally fine and in accordance with your API. I'm just not sure. Please clarify.

Reply:
In theory, yes. But for now we could say that "no". See the system-wide question above. Btw, it's probably for the first PR to not have FIE provisioning if you don't immediately connect it to resources. But we definitely then need to discuss it offline.